arkouda/numeric.py

```diff
@@ -2001,3 +2006,102 @@ def value_counts(
+
+    If lo or hi (or both) are pdarrays, the check is by pairwise elements.
+    See examples.
+
+    Parameters
+    ----------
+    pda : pdarray, int64 or float64
+        the array of values to clip
+    lo  : scalar or pdarray, int64 or float64
+        the lower value of the clipping range
+    hi  : scalar or pdarray, int64 or float64
+        the higher value of the clipping range
+    If lo or hi (or both) are pdarrays, the check is by pairwise elements.
+        See examples.
+
+    Returns
+    -------
+    arkouda.pdarrayclass.pdarray
+        A pdarray matching pda, except that element x remains x if lo <= x <= hi,
+                                                or becomes lo if x < lo,
+                                                or becomes hi if x > hi.
+
+    Examples
+    --------
+    >>> a = ak.array([1,2,3,4,5,6,7,8,9,10])
+    >>> ak.clip(a,3,8)
+    array([3,3,3,4,5,6,7,8,8,8])
+    >>> ak.clip(a,3,8.0)
+    array([3.00000000000000000 3.00000000000000000 3.00000000000000000 4.00000000000000000
+           5.00000000000000000 6.00000000000000000 7.00000000000000000 8.00000000000000000
+           8.00000000000000000 8.00000000000000000])
+    >>> ak.clip(a,None,7)
+    array([1,2,3,4,5,6,7,7,7,7])
+    >>> ak.clip(a,5,None)
+    array([5,5,5,5,5,6,7,8,9,10])
+    >>> ak.clip(a,None,None)
+    ValueError : either min or max must be supplied
+    >>> ak.clip(a,ak.array([2,2,3,3,8,8,5,5,6,6],8))
+    array([2,2,3,4,8,8,7,8,8,8])
+    >>> ak.clip(a,4,ak.array([10,9,8,7,6,5,5,5,5,5]))
+    array([4,4,4,4,5,5,5,5,5,5])
+
+    Notes
+    -----
+    Either lo or hi may be None, but not both.
+    If lo > hi, all x = hi.
+    If all inputs are int64, output is int64, but if any input is float64, output is float64.
+
+    Raises
+    ------
+    ValueError
+        Raised if both lo and hi are None
+    """
+
+    # Check that a range was actually supplied.
+
+    if lo is None and hi is None:
+        raise ValueError("Either min or max must be supplied.")
+
+    # If any of the inputs are float, then make everything float.
+    # Some type checking is needed, because scalars and pdarrays get cast differently.
+
+    dataFloat = pda.dtype == float
+    minFloat = isinstance(lo, float) or (isinstance(lo, pdarray) and lo.dtype == float)
+    maxFloat = isinstance(hi, float) or (isinstance(hi, pdarray) and hi.dtype == float)
+    forceFloat = dataFloat or minFloat or maxFloat
+    if forceFloat:
+        if not dataFloat:
+            pda = cast(pda, np.float64)
+        if lo is not None and not minFloat:
+            lo = cast(lo, np.float64) if isinstance(lo, pdarray) else float(lo)
+        if hi is not None and not maxFloat:
+            hi = cast(hi, np.float64) if isinstance(hi, pdarray) else float(hi)
+
+    # Now do the clipping.
+
+    if lo is not None:
+        pda1 = where(pda < lo, lo, pda)
+    else:
+        pda1 = pda[:]
+    if hi is not None:
+        pda1 = where(pda1 > hi, hi, pda1)
+    return pda1
```

Comment:
so this could still result in a deep copy of `pda` when it's not necessary, since we know at least one of these cases has to be true, we can safely do this:
```python
# results in a shallow copy, but it doesn't matter because it will be overwritten
pda1 = pda
if lo is not None:
    pda1 = where(pda < lo, lo, pda)
if hi is not None:
    pda1 = where(pda1 > hi, hi, pda1)
return pda1
```